In [10]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import requests
import tweepy
from tweepy import  OAuthHandler
import json 
from timeit import default_timer as timer
import os

# Gathering data

# Download twitter-archive-enhanced

In [11]:
archive=pd.read_csv('twitter-archive-enhanced.csv')

In [12]:
archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

# download tweet image

In [13]:
url='https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response=requests.get(url)


In [14]:
file_name=url.split('/')[-1]
file_name

'image-predictions.tsv'

In [15]:
if not os.path.isfile(file_name):
    with open(file_name,mode='wb') as file:
        file.write(response.content)

# import tweet data from tweeter api

In [16]:


consumer_key = '*********************************************'

consumer_secret = '*********************************************'

access_token = '*********************************************'

access_secret = '*********************************************'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth)


In [17]:
# Creating a file for the tweets' text data
errors = []
if not os.path.isfile('tweet_json.txt'):
    # create the file and write on it
    with open ('tweet_json.txt', 'w') as file:
        for tweet_id in archive['tweet_id']:
            try:
                status = api.get_status(tweet_id, wait_on_rate_limit=True,  wait_on_rate_limit_notify=True, tweet_mode = 'extended')
                json.dump(status._json, file)
                file.write('\n')
            except Exception as e:
                print("Error on tweet id {}".format(tweet_id) + ";" + str(e))
                errors.append(tweet_id)

In [18]:
df_list = []

with open('tweet_json.txt', 'r') as file:
    for line in file:
        tweet = json.loads(line)
        tweet_id = tweet['id']
        retweet_count = tweet['retweet_count']
        fav_count = tweet['favorite_count']
        df_list.append({'tweet_id':tweet_id,
                       'retweet_count': retweet_count,
                       'favorite_count': fav_count})
                                                                                                                                
api_df = pd.DataFrame(df_list)
api_df

tweet_id  retweet_count  favorite_count
0     892420643555336193           7388           35097
1     892177421306343426           5504           30383
2     891815181378084864           3636           22846
3     891689557279858688           7564           38371
4     891327558926688256           8148           36642
...                  ...            ...             ...
2326  666049248165822465             38              94
2327  666044226329800704            122             260
2328  666033412701032449             39             107
2329  666029285002620928             41             117
2330  666020888022790149            445            2344

[2331 rows x 3 columns]

# ASESS ARCHIVE

In [19]:
archive.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1013  747219827526344708                    NaN                  NaN   
1150  726224900189511680                    NaN                  NaN   
2159  669571471778410496                    NaN                  NaN   
2006  672267570918129665                    NaN                  NaN   
1159  724046343203856385                    NaN                  NaN   

                      timestamp  \
1013  2016-06-27 00:07:44 +0000   
1150  2016-04-30 01:41:23 +0000   
2159  2015-11-25 17:40:53 +0000   
2006  2015-12-03 04:14:13 +0000   
1159  2016-04-24 01:24:35 +0000   

                                                 source  \
1013  <a href="http://twitter.com/download/iphone" r...   
1150  <a href="http://twitter.com/download/iphone" r...   
2159  <a href="http://twitter.com/download/iphone" r...   
2006  <a href="http://twitter.com/download/iphone" r...   
1159  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1013  This is Atticus. He's remaining calm but his c...                  NaN   
1150  I'm getting super heckin frustrated with you a...                  NaN   
2159  This is Keith. He's had 13 DUIs. 7/10 that's t...                  NaN   
2006  When you accidentally open up the front facing...                  NaN   
1159  This is Sarge. Not even he knows what his tong...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1013                       NaN                        NaN   
1150                       NaN                        NaN   
2159                       NaN                        NaN   
2006                       NaN                        NaN   
1159                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1013  https://twitter.com/dog_rates/status/747219827...                11   
1150  https://twitter.com/dog_rates/status/726224900...                 9   
2159  https://twitter.com/dog_rates/status/669571471...                 7   
2006  https://twitter.com/dog_rates/status/672267570...                10   
1159  https://twitter.com/dog_rates/status/724046343...                10   

      rating_denominator     name doggo floofer pupper puppo  
1013                  10  Atticus  None    None   None  None  
1150                  10     None  None    None   None  None  
2159                  10    Keith  None    None   None  None  
2006                  10     None  None    None   None  None  
1159                  10    Sarge  None    None   None  None

In [20]:
archive.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [21]:
archive.rating_numerator.describe()

count    2356.000000
mean       13.126486
std        45.876648
min         0.000000
25%        10.000000
50%        11.000000
75%        12.000000
max      1776.000000
Name: rating_numerator, dtype: float64

In [22]:
archive[archive.rating_numerator<10].tweet_id

45      883482846933004288
229     848212111729840128
315     835152434251116546
387     826598799820865537
462     817502432452313088
               ...        
2351    666049248165822465
2352    666044226329800704
2353    666033412701032449
2354    666029285002620928
2355    666020888022790149
Name: tweet_id, Length: 440, dtype: int64

In [23]:
archive.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [24]:
archive.loc[(archive.tweet_id== 817502432452313088),'rating_numerator']

462    7
Name: rating_numerator, dtype: int64

In [25]:
archive.loc[462 ,'text']

"RT @dog_rates: Meet Herschel. He's slightly bigger than ur average pupper. Looks lonely. Could probably ride 7/10 would totally pet https:/…"

In [26]:
archive.rating_denominator.describe()

count    2356.000000
mean       10.455433
std         6.745237
min         0.000000
25%        10.000000
50%        10.000000
75%        10.000000
max       170.000000
Name: rating_denominator, dtype: float64

In [27]:
archive[archive.rating_denominator==0].tweet_id

313    835246439529840640
Name: tweet_id, dtype: int64

In [28]:
archive.loc[313,'text']

"@jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho"

In [29]:
archive.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [30]:
archive.name.value_counts()

None       745
a           55
Charlie     12
Lucy        11
Oliver      11
          ... 
Clarq        1
Miguel       1
Brooks       1
Nugget       1
Fiji         1
Name: name, Length: 957, dtype: int64

In [31]:
archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

# Image Predictions Assessment

In [32]:
image_df = pd.read_csv('image-predictions.tsv', sep='\t')

In [33]:
image_df.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [34]:
image_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [35]:
image_df.sample(6)

tweet_id                                          jpg_url  \
2066  890609185150312448  https://pbs.twimg.com/media/DFwUU__XcAEpyXI.jpg   
44    666781792255496192  https://pbs.twimg.com/media/CUDigRXXIAATI_H.jpg   
1967  867900495410671616  https://pbs.twimg.com/media/DAtm5MkXoAA4R6P.jpg   
488   675517828909424640  https://pbs.twimg.com/media/CV_r3v4VAAALvwg.jpg   
1036  711968124745228288  https://pbs.twimg.com/media/CeFrO3qXEAADRbd.jpg   
1090  719339463458033665  https://pbs.twimg.com/media/Cfuba6NW4AIeMHk.jpg   

      img_num                  p1   p1_conf  p1_dog            p2   p2_conf  \
2066        1       Irish_terrier  0.487574    True  Irish_setter  0.193054   
44          1   Italian_greyhound  0.618316    True    Weimaraner  0.151363   
1967        1  Labrador_retriever  0.522644    True        kuvasz  0.332461   
488         1  Scottish_deerhound  0.240591    True   groenendael  0.156916   
1036        1            espresso  0.430135   False    coffee_mug  0.418483   
1090        1    golden_retriever  0.765778    True        borzoi  0.071148   

      p2_dog                        p3   p3_conf  p3_dog  
2066    True  Chesapeake_Bay_retriever  0.118184    True  
44      True                    vizsla  0.085989    True  
1967    True                 dalmatian  0.032008    True  
488     True     flat-coated_retriever  0.090899    True  
1036   False                       cup  0.088391   False  
1090    True                  Leonberg  0.070371    True

In [36]:
image_df.loc[image_df.p2=='cup'].tweet_id

1851    840370681858686976
1887    847971574464610304
Name: tweet_id, dtype: int64

In [37]:
image_df.loc[1851,'jpg_url']

'https://pbs.twimg.com/media/C6mYrK0UwAANhep.jpg'

In [38]:
from IPython.display import Image
Image(url ='https://pbs.twimg.com/media/C6mYrK0UwAANhep.jpg')

# Tweet JSON  txt Assessment

In [39]:
api_df.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           7388           35097
1  892177421306343426           5504           30383
2  891815181378084864           3636           22846
3  891689557279858688           7564           38371
4  891327558926688256           8148           36642

In [40]:
api_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2331 non-null   int64
 1   retweet_count   2331 non-null   int64
 2   favorite_count  2331 non-null   int64
dtypes: int64(3)
memory usage: 54.8 KB


In [41]:
api_df.sample(10)

tweet_id  retweet_count  favorite_count
2141  669363888236994561            209             581
92    874057562936811520           3517           20840
1931  673612854080196609            668            1501
2140  669367896104181761            140             426
760   775842724423557120           2587           11471
2133  669573570759163904            127             404
1979  672466075045466113            485            1260
1907  674024893172875264           1137            1664
515   808001312164028416           3478           12355
1970  672594978741354496            567            1240

# Tidiness

In [42]:
#all data are related but the are sperated

# Twitter Archive

In [43]:
#dog types seprated into 4 columns


# Image_url

In [44]:
#p1,p2,p3 should be renamed to more descriptive names.

# Tweet JSON

# Quality

# Twitter Archive

1-timestamp shoud be datetime

2-Numerator and denominator wrong data type and extracting incorrect fractions

3-names are invalid(none,a,an)

4-invalid tweet_id datatype(string not integer)

5-23 rating denominator !=10

6-raw 313 rating denominator = 0

7-more rating numerator less than 10

8-delete there are 181 retweeted_id





# Tweet Image

1_missing photo (2075 instead 2356)
2_some p names start with uppercase while other start with lowercase
3_delete uderscores in p1,p2,p3
4_dog_type should be categroy
5_rename p names to be  more descriptive names.

# Tweet Api

missing entries (2331 instead of 2356)

# Cleaning Data

# copy for data

In [45]:
archive_clean=archive.copy()
image_clean=image_df.copy()
api_clean=api_df.copy()

In [46]:
archive_clean.sample(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2263  667544320556335104                    NaN                  NaN   

                      timestamp  \
2263  2015-11-20 03:25:43 +0000   

                                                 source  \
2263  <a href="http://twitter.com" rel="nofollow">Tw...   

                                                   text  retweeted_status_id  \
2263  This is Kial. Kial is either wearing a cape, w...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2263                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2263  https://twitter.com/dog_rates/status/667544320...                10   

      rating_denominator  name doggo floofer pupper puppo  
2263                  10  Kial  None    None   None  None

**define:
collect four column in one column called dog_type

code

In [47]:
archive_clean['dog_type']=archive_clean['text'].str.extract('(doggo|floofer|pupper|puppo)')
archive_clean.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo dog_type  
0                  10   Phineas  None    None   None  None      NaN  
1                  10     Tilly  None    None   None  None      NaN  
2                  10    Archie  None    None   None  None      NaN  
3                  10     Darla  None    None   None  None      NaN  
4                  10  Franklin  None    None   None  None      NaN

In [48]:
archive_clean=archive_clean.drop(columns=['doggo','floofer','pupper','puppo'])

 Test

In [49]:
archive_clean

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

In [50]:
archive_clean.dog_type.value_counts()

pupper     265
doggo       93
puppo       37
floofer      4
Name: dog_type, dtype: int64

# all data are related but the are sperated

#Define:
merg all dataframe in one based on tweet_id

Code

In [51]:
df_new=pd.merge(archive_clean,api_clean,on='tweet_id',how='left')
df_new=pd.merge(df_new,image_clean,on='tweet_id',how='left')

In [52]:
Test

NameError: name 'Test' is not defined

In [ ]:
df_new.info()

In [ ]:
pd.set_option('display.max_columns', None)

In [ ]:

df_new.head()

# p1,p2,p3 should be renamed to more descriptive names.

# cleaning quality 

# timestamp shoud be datetime

Define:timestamp is datetime type so i convert into that

Code

In [ ]:
df_new['timestamp'] = pd.to_datetime(df_new.timestamp)

In [ ]:
Test

In [ ]:
df_new.info()

# invalid tweet_id datatype(string not integer)

DEfine:
    convert nvalid tweet_id datatype to string

In [ ]:
Code

In [ ]:
df_new['tweet_id'] =df_new .tweet_id.astype('object')

In [ ]:
Test

In [ ]:
df_new.info()

# delete there are 181 retweeted_id

Define
#delete al rows related with (retweeted_status_id , retweeted_status_user_id ,retweeted_status_timestamp)

Code and Test

In [ ]:
df_new=df_new[df_new.retweeted_status_id.isnull()]
df_new.info()

In [ ]:
df_new=df_new.drop(columns=['retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp'])


In [ ]:
df_new.head()

# missing photo from some ids

DEfine:
    delete rows with missing data

Code

In [ ]:
df_new=df_new[df_new.jpg_url.notnull()]

In [ ]:
Test

In [ ]:
df_new.info()

# uderscores in p1,p2,p3

delete underscores in p1 ,p2 ,p3

In [ ]:
Code

In [ ]:
df_new.p1=df_new.p1.str.replace('_',' ')
df_new.p2=df_new.p2.str.replace('_',' ')
df_new.p3=df_new.p3.str.replace('_',' ')

In [ ]:
Test

In [ ]:
df_new.p1.head(5)

In [ ]:
df_new.p1.value_counts()

# Some p names start lowercase other lowercase

In [ ]:
Define:
convert lowercase to uppercase

In [ ]:
Code

In [ ]:
df_new.p1=df_new.p1.str.title()
df_new.p2=df_new.p2.str.title()
df_new.p3=df_new.p3.str.title()

In [ ]:
Test

In [ ]:
df_new.p1.value_counts()

# Missing enteries retweet_count

Define:
    delete rows without retweet_count entry

Code:

In [ ]:
df_new=df_new[df_new.retweet_count.notnull()]

In [ ]:
sum(df_new.retweet_count.isnull())

Test:

In [ ]:
df_new.info()

In [ ]:
df_new.head()

# dog_type should be category

Define dog_type shouhd convert into catergory type

Code

In [ ]:
df_new['dog_type'] = df_new.dog_type.astype('category')

Test

In [ ]:
df_new.info()

# Missing enteries favorite_count


In [ ]:
Define: delete rows without favorite_count entry

In [ ]:
Code:

In [ ]:
df_new=df_new[df_new.favorite_count.notnull()]

In [ ]:
Test:

In [ ]:
df_new.info()

# Delete in_reply_to_status_id,in_reply_to_user_id


Define:Delete in_reply_to_status_id,in_reply_to_user_id
    

In [ ]:
Code and Test

In [ ]:
df_new=df_new.drop(columns=['in_reply_to_status_id','in_reply_to_user_id'])
df_new.info()

# Rename p names to be  more descriptive names.

In [ ]:
Define:
    we will rename p names column to be more more descriptive names
    

Code

In [ ]:
# Renaming the dataset columns
cols = ['tweet_id','timestamp' ,'source', 'text' ,
      'expanded_urls ' , 'rating_numerator ','rating_denominator ',
        'name','dog_type',
        'retweet_count','favorite_count','jpg_url','img_num'
        ,'prediction_1'
        , 'confidence_1', 
        'breed_1',
       'prediction_2', 'confidence_2', 'breed_2',
       'prediction_3', 'confidence_3', 'breed_3']
df_new.columns = cols


Test:

In [ ]:
df_new.info()

In [ ]:
df_new


# Stroing Data

In [ ]:
df_new.to_csv('twitter_archive_master.csv')

# Data Analysis Ana Visulization

# the precentage of dog_type

In [ ]:
dog_type=df_new.dog_type.value_counts(    )

In [ ]:
dog_type

In [ ]:
plt.pie(dog_type,
labels=['pupper','doggo','puppo ','floofer'],   
       shadow=True,
       autopct='%1.1f%%',
       explode=(0.1,0.1,0.3,0.3))
plt.title('the percentage of dog_type')

In [ ]:
the insights:
   1_ pupper has the highest percentage
   2_ floofer has the lowest percentage

# the relation between retweet_count and favorite_count


In [ ]:
plt.scatter(df_new.retweet_count,df_new.favorite_count  )
plt.title('the relation between retweet_count and favorite_count')
plt.xlabel('retweet_count')
plt.ylabel('favorite_count')

In [ ]:
the insights:
    1_retweet count less than favrouite count
    2_where retweet count increase the favorite count increas(linear relation)

In [ ]:
avg_retweet_count_dog_stage = df_new.groupby('dog_type').retweet_count.mean()

# the index with stand for the x axis in plotting 
avg_retweet_count_dog_stage.index

avg_favorite_count_dog_stage = df_new.groupby('dog_type').favorite_count.mean()  

# the mean of favorite_count for dog_type

In [ ]:
locations=[1,2,3,4]
labels=['pupper','doggo','puppo ','floofer']
plt.bar(locations,avg_favorite_count_dog_stage,tick_label=labels)
plt.title('the mean  favorite_count for dog_type ')
plt.xlabel('dog_type')
plt.ylabel('favorite_count')

# the mean of retweet_count for dog_type

In [ ]:
locations=[1,2,3,4]
labels=['pupper','doggo','puppo ','floofer']
plt.bar(locations,avg_retweet_count_dog_stage,tick_label=labels)
plt.title('the mean  retweet_count for dog_type ')
plt.xlabel('dog_type')
plt.ylabel('retweet_count')

puppo has the less likes and retweet